src/yuzu/game_list.cpp

```diff
@@ -344,19 +470,86 @@ void GameList::PopupContextMenu(const QPoint& menu_location) {
-    connect(dump_romfs, &QAction::triggered, [&]() { emit DumpRomFSRequested(program_id, path); });
-    connect(copy_tid, &QAction::triggered, [&]() { emit CopyTIDRequested(program_id); });
-    connect(navigate_to_gamedb_entry, &QAction::triggered,
-            [&]() { emit NavigateToGamedbEntryRequested(program_id, compatibility_list); });
-    connect(properties, &QAction::triggered, [&]() { emit OpenPerGameGeneralRequested(path); });
+            [this, program_id]() { emit OpenTransferableShaderCacheRequested(program_id); });
+    connect(dump_romfs, &QAction::triggered,
+            [this, program_id, path]() { emit DumpRomFSRequested(program_id, path); });
+    connect(copy_tid, &QAction::triggered,
+            [this, program_id]() { emit CopyTIDRequested(program_id); });
+    connect(navigate_to_gamedb_entry, &QAction::triggered, [this, program_id]() {
+        emit NavigateToGamedbEntryRequested(program_id, compatibility_list);
+    });
+    connect(properties, &QAction::triggered,
+            [this, path]() { emit OpenPerGameGeneralRequested(path); });
+};
+
+void GameList::AddCustomDirPopup(QMenu& context_menu, QModelIndex selected) {
+    UISettings::GameDir& game_dir =
+        *selected.data(GameListDir::GameDirRole).value<UISettings::GameDir*>();
+
+    QAction* deep_scan = context_menu.addAction(tr("Scan Subfolders"));
+    QAction* delete_dir = context_menu.addAction(tr("Remove Game Directory"));
+
+    deep_scan->setCheckable(true);
+    deep_scan->setChecked(game_dir.deep_scan);
+
+    connect(deep_scan, &QAction::triggered, [this, &game_dir] {
+        game_dir.deep_scan = !game_dir.deep_scan;
+        PopulateAsync(UISettings::values.game_dirs);
+    });
+    connect(delete_dir, &QAction::triggered, [this, &game_dir, selected] {
+        UISettings::values.game_dirs.removeOne(game_dir);
+        item_model->invisibleRootItem()->removeRow(selected.row());
+    });
+}
 
-    context_menu.exec(tree_view->viewport()->mapToGlobal(menu_location));
+void GameList::AddPermDirPopup(QMenu& context_menu, QModelIndex selected) {
+    UISettings::GameDir& game_dir =
+        *selected.data(GameListDir::GameDirRole).value<UISettings::GameDir*>();
+
+    QAction* move_up = context_menu.addAction(tr(u8"\U000025b2 Move Up"));
+    QAction* move_down = context_menu.addAction(tr(u8"\U000025bc Move Down "));
+    QAction* open_directory_location = context_menu.addAction(tr("Open Directory Location"));
+
+    int row = selected.row();
+
+    move_up->setEnabled(row > 0);
+    move_down->setEnabled(row < item_model->rowCount() - 2);
+
+    connect(move_up, &QAction::triggered, [this, selected, row, &game_dir] {
+        // find the indices of the items in settings and swap them
+        UISettings::values.game_dirs.swap(
+            UISettings::values.game_dirs.indexOf(game_dir),
+            UISettings::values.game_dirs.indexOf(*selected.sibling(selected.row() - 1, 0)
```

Comment:
This can likely replace the selected query with:

```suggestion
            UISettings::values.game_dirs.indexOf(*selected.sibling(row - 1, 0)
```